Fix for not marking peer as low quality too soon if there is only one peer connected

ManiBAJPAI22 opened the pull request:
Fixes  #6803 

_Please choose one of the keywords above to refer to the issue this PR solves followed by the issue number (e.g. Fixes #000). If no issue number, remove the line. Also, remove everything marked optional that is not applicable. Remove this note after reading._

## Changes

- _List the changes_

## Types of changes

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

## Testing

#### Requires testing

- [ ] Yes
- [x] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing

_Optional. Remove if not applicable._

## Documentation

#### Requires documentation update

- [ ] Yes
- [x] No

_If yes, link the PR to the docs update or the issue with the details labeled `docs`. Remove if not applicable._

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

_If yes, fill in the details here. Remove if not applicable._

## Remarks

_If the number of consecutive failures from the current peer (peerLastFailures) exceeds the AllowedInvalidResponses threshold, and all the recent failures are from this peer ```(allLastFailures == peerLastFailures)```, the peer is marked as LesserQuality.

In a scenario where there is only one peer connected, this condition will be met quickly, leading to the peer being marked as ```LesserQuality``` even though there are no other peers to compare with.

To address this issue and improve the behavior in situations with a single peer, we can modify the logic to update the pivot and retry with the same peer before marking it as ```LesserQuality```, when the current peer is the only one with failures, instead of immediately marking it as ```LesserQuality```, the code will update the pivot, clear the result log, and log a message indicating that the pivot has been updated and the peer will be retried.

The ```SyncResponseHandlingResult.OK``` is returned, allowing the sync process to retry with the same peer after updating the pivot.
If there are other peers with failures ```(allLastFailures != peerLastFailures)```, then the original logic of marking the current peer as LesserQuality is preserved.

This change should improve the behavior in scenarios with a single peer by giving the peer another chance after updating the pivot, before marking it as ```LesserQuality```. _

ManiBAJPAI22 committed:
Fix low quality peer issue (one peer)

asdacap commented on src/Nethermind/Nethermind.Synchronization/SnapSync/SnapSyncFeed.cs:
```diff
@@ -162,15 +164,29 @@ public SyncResponseHandlingResult AnalyzeResponsePerPeer(AddRangeResult result,
-                                    }
+                                      peerLastFailures++;
+
+                                               if (peerLastFailures > AllowedInvalidResponses)
```
Please cleanup indentation

ManiBAJPAI22 replied:
Sure !!

ManiBAJPAI22 committed:
Update SnapSyncFeed.cs

Indentations updated

ManiBAJPAI22 committed:
Merge pull request #1 from ManiBAJPAI22/ManiBAJPAI22-branch-1

Update SnapSyncFeed.cs

ManiBAJPAI22 committed:
Merge branch 'NethermindEth:master' into master

ManiBAJPAI22 commented:
Indentations managed 
@asdacap kindly review it.

benaadams commented on src/Nethermind/Nethermind.Synchronization/SnapSync/SnapSyncFeed.cs:
```diff
@@ -168,18 +168,29 @@ public SyncResponseHandlingResult AnalyzeResponsePerPeer(AddRangeResult result,
+                                        
+                                        return SyncResponseHandlingResult.OK;
+                                    }
+                                    else
```
Won't this mark peer as lesser quality unless all failures were that peer? (i.e. if its less than all, it will also mark as Lesser)

ManiBAJPAI22 committed:
Update/2 with new boolean variable peerIsOnlyFailure SnapSyncFeed.cs

The key changes are:

Introduced a new boolean variable peerIsOnlyFailure and initialized it to true.
While iterating through _resultLog, if a failure is encountered from a peer other than the current peer, peerIsOnlyFailure is set to false.
When checking if peerLastFailures is greater than AllowedInvalidResponses, the code first checks the value of peerIsOnlyFailure.

If peerIsOnlyFailure is true, it means the current peer is solely responsible for all the failures, so it updates the pivot and returns SyncResponseHandlingResult.OK.
If peerIsOnlyFailure is false, it means there are failures from other peers as well, so it marks the current peer as SyncResponseHandlingResult.LesserQuality.

ManiBAJPAI22 committed:
Merge pull request #2 from ManiBAJPAI22/ManiBAJPAI22-branch-1-1

Update/2 with new boolean variable peerIsOnlyFailure SnapSyncFeed.cs

ManiBAJPAI22 commented:
@benaadams, @asdacap,

I've updated the code to address the concern regarding the peer quality evaluation logic. I nearly missed that detail, so thank you for pointing it out.

ManiBAJPAI22 committed:
Merge branch 'NethermindEth:master' into master

ManiBAJPAI22 commented:
@benaadams @asdacap 
Any other thing that I need to do ?

asdacap commented on src/Nethermind/Nethermind.Synchronization/SnapSync/SnapSyncFeed.cs:
```diff
@@ -110,90 +110,93 @@ public override SyncResponseHandlingResult HandleResponse(SnapSyncBatch? batch,
         }
 
         public SyncResponseHandlingResult AnalyzeResponsePerPeer(AddRangeResult result, PeerInfo peer)
+{
```
Indentation is still wrong.

ManiBAJPAI22 replied:
Sorry for the delayed response. 
I guess now Indentations are managed.

ManiBAJPAI22 committed:
Update SnapSyncFeed.cs

Indentations managed

ManiBAJPAI22 committed:
Merge pull request #3 from ManiBAJPAI22/ManiBAJPAI22-patch-3

Update SnapSyncFeed.cs

benaadams committed:
Merge branch 'master' into master

asdacap reviewed (approved):
LGTM.

asdacap commented:
Should put a unit tests for this case though.

ManiBAJPAI22 commented:
Yeah checking on it !

stdevMac commented:
Closing — the unit test asdacap requested in Sept 2024 was never added, and the branch is ~600 days stale. If #6803 still reproduces, please reopen with a focused fix and tests.
